### PR TITLE
Add tracestate handling

### PR DESCRIPTION
**Pull request by rdooley**
* set, get, del
* yangxikun/opentelemetry-lua#50

### CODE REVIEW

File: lib/opentelemetry/trace/tracestate.lua [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+-- Set the key value pair for the tracestate
+--
+-- @return              tracestate
+--------------------------------------------------------------------------------
+function _M.set(self, key, value)
+    if not validate_member_key(key) then
+        return self
+    end
+    self:del(key)
+    if not validate_member_value(value) then
```
Comment (plantfansam) [inline lib/opentelemetry/trace/tracestate.lua:97]: validate value before deleting?

File: lib/opentelemetry/trace/tracestate.lua [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        return self
+    end
+    if #self >= _M.MAX_ENTRIES then
```
Comment (plantfansam) [inline lib/opentelemetry/trace/tracestate.lua:100]: Thoughts about logging a warning here? This is really bad, if it happens. Popping stuff off the state 😬

File: lib/opentelemetry/trace/tracestate.lua [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    end
+
+    local new_tracestate = {}
+    local members_count = 0
+    for _, item in ipairs(tracestate) do
+        for member in string.gmatch(item, "([^,]+)") do
+            if member ~= "" then
+                local start_pos, end_pos = string.find(member, "=", 1, true)
+                if not start_pos or start_pos == 1 then
+                    return _M.new({})
```
[the remark this answers is not in the document]
Comment (rdooley) [reply lib/opentelemetry/trace/tracestate.lua:63]: yes we should log
re returning an empty tracestate, that is what the [go sdk is doing](https://github.com/open-telemetry/opentelemetry-go/blob/ff51300a2af2581f91d7427ad50d1939f8e99cdb/trace/tracestate.go#L102)

File: lib/opentelemetry/trace/tracestate.lua [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+--------------------------------------------------------------------------------
+-- Parse tracestate header into a tracestate
+--
+-- @return              tracestate
+--------------------------------------------------------------------------------
+function _M.parse_tracestate(tracestate)
+    if not tracestate then
+        return _M.new({})
+    end
+    if type(tracestate) == "string" then
```
[the remark this answers is not in the document]
Comment (rdooley) [reply lib/opentelemetry/trace/tracestate.lua:52]: fair call out, this was a `yypp` development thing

File: lib/opentelemetry/trace/span_context.lua [hunk 5]
```diff
@@ -12,7 +13,7 @@ function _M.new(tid, sid, trace_flags, trace_state, remote)
         trace_id    = tid,
         span_id     = sid,
         trace_flags = trace_flags,
-        trace_state = trace_state,
+        trace_state = trace_state or tracestate.parse_tracestate(trace_state),
```
[the remark this answers is not in the document]
Comment (rdooley) [reply lib/opentelemetry/trace/span_context.lua:16]: i think i just had this earlier to make a test not break and should clean up now

File: lib/opentelemetry/trace/tracestate.lua [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+function _M.set(self, key, value)
+    if not validate_member_key(key) then
+        return self
+    end
+    self:del(key)
+    if not validate_member_value(value) then
+        return self
+    end
+    if #self >= _M.MAX_ENTRIES then
+        table.remove(self)
```
[the remark this answers is not in the document]
Comment (rdooley) [reply lib/opentelemetry/trace/tracestate.lua:101]: I did add a ` it("max len is respected", function()` test can you elaborate on what other logic the test should cover?

File: lib/opentelemetry/trace/span_context.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (rdooley) [reply lib/opentelemetry/trace/span_context.lua:16]: no that's exactly why we have this, if trace_state is nil `parse_tracestate` will give us an empty one to use

File: lib/opentelemetry/trace/tracestate.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (rdooley) [reply lib/opentelemetry/trace/tracestate.lua:52]: also needed for the e2e tests

File: lib/opentelemetry/trace/tracestate.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (plantfansam) [reply lib/opentelemetry/trace/tracestate.lua:101]: Now I'm second guessing myself. The scenario I'm thinking is as follows:

Someone finds a blog post saying `table.remove` is bad and updates `del()` to do:

```
- `table.remove(self, index)
+ self[index] = nil
```

Then, all of a sudden #self doesn't actually represent the number of entries in the table. Potentially I'm being paranoid.

File: lib/opentelemetry/trace/tracestate.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (rdooley) [reply lib/opentelemetry/trace/tracestate.lua:52]: we are missing a validation in this function for duplicate keys in tracestate which should trigger an error

[the remark this answers is not in the document]
Comment (rdooley) [reply lib/opentelemetry/trace/tracestate.lua:52]: actually that fails the w3c testcase `test_tracestate_duplicated_keys` which expects to get both or just one of the duplicated keys back. Interesting choice by go sdk we will not be duplicating here

File: lib/opentelemetry/trace/tracestate.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (plantfansam) [reply lib/opentelemetry/trace/tracestate.lua:66]: OK. I'm fine with matching go.

File: lib/opentelemetry/trace/span_context.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (plantfansam) [reply lib/opentelemetry/trace/span_context.lua:16]: Purely stylistic, but might go with `trace_state or trace_state.new({})` or whatever.

File: lib/opentelemetry/trace/tracestate.lua [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    return _M.new(new_tracestate)
+end
+
+--------------------------------------------------------------------------------
+-- Set the key value pair for the tracestate
+--
+-- @return              tracestate
+--------------------------------------------------------------------------------
+function _M.set(self, key, value)
+    if not validate_member_key(key) then
```
Comment (plantfansam) [inline lib/opentelemetry/trace/tracestate.lua:97]: Might be worth an info log if attempting to set invalid key/value 🤷

File: lib/opentelemetry/trace/span_context.lua (hunk 5 above)
[the remark this answers is not in the document]
Comment (yangxikun) [reply lib/opentelemetry/trace/span_context.lua:16]: I agree with `trace_state or trace_state.new({})`, these mean the `trace_state` param can be nil or `tracestate` table.
`trace_state or tracestate.parse_tracestate(trace_state)` means we can pass `nil` or `tracestate` table or tracestate string.
I think func param should have one certainly type.
Also, parse tracestate is the responsibility of propagation.

File: lib/opentelemetry/trace/tracestate.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (plantfansam) [reply lib/opentelemetry/trace/tracestate.lua:53]: I get it now. If there are multiple `tracestate` headers on a request, `ngx.req.get_headers()["tracestate"]` [returns a table](https://github.com/openresty/lua-nginx-module#ngxreqget_headers) 😅.  Do you mind adding a comment between line 51 and 52 like "if there's only one tracestate header, it comes in as a string; if there's more than one, they come in as a table"?

**Review by plantfansam — APPROVED** [review]
I had a couple nits but I think this is good to merge. @yangxikun what do you think?

**Pull request merged**